firewall.go

```diff
@@ -109,6 +109,53 @@ func (id *LayerID) IsZero() bool {
 	return id == nil || *id == LayerID{}
 }
 
+// ConditionFlag represents special conditions that can be tested.
+type ConditionFlag uint32 // do not change type, used in C calls
+
+const (
+	ConditionFlagIsLoopback             ConditionFlag = 0x00000001
+	ConditionFlagIsIPSecSecured         ConditionFlag = 0x00000002
+	ConditionFlagIsReauthorize          ConditionFlag = 0x00000004
+	ConditionFlagIsWildcardBind         ConditionFlag = 0x00000008
+	ConditionFlagIsRawEndpoint          ConditionFlag = 0x00000010
+	ConditionFlagIsFragmant             ConditionFlag = 0x00000020
+	ConditionFlagIsFragmantGroup        ConditionFlag = 0x00000040
+	ConditionFlagIsIPSecNATTReclassify  ConditionFlag = 0x00000080
+	ConditionFlagIsRequiresALEClassify  ConditionFlag = 0x00000100
+	ConditionFlagIsImplicitBind         ConditionFlag = 0x00000200
+	ConditionFlagIsReassembled          ConditionFlag = 0x00000400
+	ConditionFlagIsNameAppSpecified     ConditionFlag = 0x00004000
+	ConditionFlagIsPromiscuous          ConditionFlag = 0x00008000
+	ConditionFlagIsAuthFW               ConditionFlag = 0x00010000
+	ConditionFlagIsReclassify           ConditionFlag = 0x00020000
+	ConditionFlagIsOutboundPassThru     ConditionFlag = 0x00040000
+	ConditionFlagIsInboundPassThru      ConditionFlag = 0x00080000
+	ConditionFlagIsConnectionRedirected ConditionFlag = 0x00100000
+)
+
+// IPProto represents the protocol being used.
+type IPProto uint8 // do not change type, used in C calls
+
+// From: https://docs.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-socket
+const (
+	IPProtoICMP   IPProto = 1
+	IPProtoICMPV6 IPProto = 58
+	IPProtoTCP    IPProto = 6
+	IPProtoUDP    IPProto = 17
+)
+
+// AppID returns the application ID associated with the provided file.
+func AppID(file string) (string, error) {
+	var a arena
+	defer a.Dispose()
+	fileBytes, _ := toBytesFromString(&a, file)
+	appID := (*fwpByteBlob)(a.Alloc(unsafe.Sizeof(fwpByteBlob{})))
+	if err := fwpmGetAppIdFromFileName0(fileBytes, &appID); err != nil {
```

Comment:
This call isn't right, the function returns allocated memory, and you need to free it with fwpmFreeMemory0. See https://docs.microsoft.com/en-us/windows/win32/api/fwpmu/nf-fwpmu-fwpmgetappidfromfilename0 for how it's used in C++.

Should be something like:

var blob *fwpByteBlob
fwpmGetAppIdFromFileName0(fileBytes, &blob)
defer fwpmFreeMemory0(&blob)
...

Reply:
Thanks for catching this!